SFR-653 Update Gutenberg EPUB filename extensions

mwbenowitz opened the pull request:
Project Gutenberg stores its EPUB files with custom extensions `.images` and `.noimages` to distinguish between multiple copies of the same instance from their collection. While epub.js was able to read files in this format the webpub reader is not, nor is iBooks.

This transforms these filenames from the format `00000.epub.images` to `00000_images.epub` and does not transform and non-standard EPUB file names received from Gutenberg or other filenames received from other sources (as long as they end in .epub)

mwbenowitz committed:
SFR-653 Update Gutenberg EPUB filename extensions

Project Gutenberg stores its EPUB files with custom extensions `.images` and `.noimages` to distinguish between multiple copies of the same instance from their collection. While epub.js was able to read files in this format the webpub reader is not, nor is iBooks.

This transforms these filenames from the format `00000.epub.images` to `00000_images.epub` and does not transform and non-standard EPUB file names received from Gutenberg or other filenames received from other sources (as long as they end in .epub)

seanredmond commented on src/helpers/fileNameParser.js:
```diff
@@ -0,0 +1,7 @@
+const fileNameRegex = /([0-9]+)\.epub\.((?:no)?images)$/
```
The first group could just be anything, `(^.+)`. I would at least add the `^` because if any Gutenberg file ever had something _before_ the numbers this regex would cut it off.

mwbenowitz replied:
As we discussed the intention was to extract that "filename" from the URL. But the approach is slightly different and more general now in the new commit.

mwbenowitz committed:
SFR-563 Improve Regex filename parsing/validation

This makes several improvements to how filenames are parsed by this function. First, it merges to separate regex steps into a single method in the `helpers` module. This is a better assignment of responsibilities and allows the function to utilize a single regex for both validation and parsing. Second, this makes the regex applied a more general pattern. Previously the check was limited to Project Gutenberg ePub files, which has been replaced with a regex that should generally validated `.epub` files while retaining the parsing necessary for Gutenberg's odd naming conventions.

Finally this adds some in-line documentation of the regex and the new method, which hopefully makes this more clear.

Additional tests have been updated and some functionality removed from `index.js` as this is now managed by the helper function.

mwbenowitz commented:
Added a new commit that makes this functionality more general and able to apply for `.epub` files from all sources, which will be important moving forward. Also included some documentation to make the fairly opaque helper function more clear.

hokei commented on test/fileNameParser.test.js:
```diff
@@ -0,0 +1,35 @@
+chai.use(chaiAsPromised)
+const { expect } = chai
+
+describe('helpers/fileNameParser', () => {
```
maybe add one more test having it reject non-epub files with the error code?

hokei commented on test/handler.test.js:
```diff
@@ -176,14 +172,39 @@ describe('Handlers [index.js]', () => {
 
     it('should throw LambdaError if regex match fails', () => {
+      testData.data = {
+        url: 'http://www.gutenberg.org/ebooks/bad.epub.images',
```
Why would the regex match fail on this?  Is it because it's re-assigned in line 208?  (it's simpler to assign the /notReal url here?)

mwbenowitz replied:
Thanks for catching this, I'm not sure why I had the test that way. You're correct that it should just start with the `/notReal` URL. (And you're right that the `bad.epub.images` would actually be a valid URL)

hokei commented on test/fileNameParser.test.js:
```diff
@@ -0,0 +1,35 @@
+chai.use(chaiAsPromised)
+const { expect } = chai
+
+describe('helpers/fileNameParser', () => {
```
Perhaps add a test rejecting a filename with no epub in it at all?

mwbenowitz replied:
Good idea, adding that test now.

mwbenowitz committed:
SFR-563 Update Test Suite

An outdated test was confusingly setting a valid URL in a test designed to throw an error before overwriting it, which doesn't make sense as the URL will never be changed within the function. This simply sets the invalid URL for the test. The helper function's test file also did not include a test to raise the `regex-failure` error, and that has been added to account for all possible outputs of the function.
